File: codeexamples/solid/better/CheckoutServiceMainApplication.java
package better;

import java.util.Map;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * ===============================
 * 1. Abstraction (DIP)
 * ===============================
 */
interface PaymentProcessor {
    void pay(double amount);
}

/**
 * ===============================
 * 2. Concrete Implementations (SRP)
 * ===============================
 */
class StripePaymentProcessor implements PaymentProcessor {

    private static final Logger logger = Logger.getLogger(StripePaymentProcessor.class.getName());
    private static final String PAYMENT_MESSAGE = "Paid $ %s using Stripe";

    @Override
    public void pay(double amount) {
        if (logger.isLoggable(Level.INFO)) {
            logger.info(String.format(PAYMENT_MESSAGE, amount));
        }
    }
}

class RazorpayPaymentProcessor implements PaymentProcessor {

    private static final Logger logger = Logger.getLogger(RazorpayPaymentProcessor.class.getName());
    private static final String PAYMENT_MESSAGE = "Paid $ %s using Razorpay";

    @Override
    public void pay(double amount) {
        if (logger.isLoggable(Level.INFO)) {
            logger.info(String.format(PAYMENT_MESSAGE, amount));
        }
    }
}

class PayPalPaymentProcessor implements PaymentProcessor {

    private static final Logger logger = Logger.getLogger(PayPalPaymentProcessor.class.getName());
    private static final String PAYMENT_MESSAGE = "Paid $%s using PayPal";

    @Override
    public void pay(double amount) {
        if (logger.isLoggable(Level.INFO)) {
            logger.info(String.format(PAYMENT_MESSAGE, amount));
        }
    }
}

enum PaymentProvider {
    STRIPE,
    RAZORPAY,
    PAYPAL
}

/**
 * ===============================
 * 3. Factory (OCP)
 * ===============================
 */
class PaymentProcessorFactory {

    private final Map<PaymentProvider, PaymentProcessor> processors;

    public PaymentProcessorFactory(Map<PaymentProvider, PaymentProcessor> processors) {
        this.processors = processors;
    }

    public PaymentProcessor get(PaymentProvider provider) {

        if (provider == null) {
            throw new IllegalArgumentException("Payment provider must be specified");
        }

        PaymentProcessor processor = processors.get(provider);

        if (processor == null) {
            throw new IllegalArgumentException(
                    "Unsupported payment provider: " + provider);
        }

        return processor;
    }
}

/**
 * ===============================
 * 4. Business Logic (SRP + DIP)
 * ===============================
 */
class CheckoutService {

    private static final Logger logger = Logger.getLogger(CheckoutService.class.getName());

    private final PaymentProcessorFactory factory;

    public CheckoutService(PaymentProcessorFactory factory) {
        this.factory = factory;
    }

    /**
     * Executes checkout using selected payment provider.
     *
     * @param amount   payment amount (must be > 0)
     * @param provider payment provider name
     */
    public void checkout(double amount, PaymentProvider provider) {

        if (amount <= 0) {
            throw new IllegalArgumentException("Amount must be greater than zero");
        }

        if (logger.isLoggable(Level.INFO)) {
            logger.log(Level.INFO, "Starting checkout with provider: {0}", provider);
        }

        PaymentProcessor processor = factory.get(provider);
        processor.pay(amount);
    }
}

public class CheckoutServiceMainApplication {
    public static void main(String[] args) {
        PaymentProcessorFactory factory = new PaymentProcessorFactory(
                Map.of(
                        PaymentProvider.STRIPE, new StripePaymentProcessor(),
                        PaymentProvider.RAZORPAY, new RazorpayPaymentProcessor(),
                        PaymentProvider.PAYPAL, new PayPalPaymentProcessor()));

        CheckoutService checkoutService = new CheckoutService(factory);

        checkoutService.checkout(1500.0, PaymentProvider.STRIPE);
    }
}

/*
 * ===============================
 * Output    
 * ===============================
 * 
s2tha@thalasi-windows MINGW64 /d/git/low-level-design-practices/codeexamples/solid/better (main)
$ java -cp .. better.CheckoutService
Feb 07, 2026 10:18:14 AM better.CheckoutService checkout
INFO: Starting checkout with provider: STRIPE
Feb 07, 2026 10:18:14 AM better.StripePaymentProcessor pay
INFO: Paid ?1500.0 using Stripe

s2tha@thalasi-windows MINGW64 /d/git/low-level-design-practices/codeexamples/solid/better (main)
$ javac CheckoutServiceMainApplication.java

s2tha@thalasi-windows MINGW64 /d/git/low-level-design-practices/codeexamples/solid/better (main)
$ javac CheckoutServiceMainApplication.java

s2tha@thalasi-windows MINGW64 /d/git/low-level-design-practices/codeexamples/solid/better (main)
$ java -cp .. better.CheckoutServiceMainApplication
Feb 07, 2026 10:30:29 AM better.CheckoutService checkout
INFO: Starting checkout with provider: STRIPE
Feb 07, 2026 10:30:29 AM better.StripePaymentProcessor pay
INFO: Paid $ 1500.0 using Stripe

s2tha@thalasi-windows MINGW64 /d/git/low-level-design-practices/codeexamples/solid/better (main)
$

==============================
5. SOLID Principles 
==============================

 * 
 * | Principle | Status | Why |
 * | ----------- | ------ | --------------------------------------- |
 * | **S – SRP** | Y | Checkout ≠ provider creation |
 * | **O – OCP** | Y | Add provider without modifying Checkout |
 * | **L – LSP** | Y | All processors interchangeable |
 * | **I – ISP** | Y | Small focused interface |
 * | **D – DIP** | Y | Depends on abstraction |
 * 
 */

File: codeexamples/solid/better/CheckoutServiceMainApplication.py
# package better;

from abc import ABC, abstractmethod
from enum import Enum
from typing import Dict
import logging

# ===============================
# Logging Configuration
# ===============================
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

"""
===============================
1. Abstraction (DIP)
===============================
"""
class PaymentProcessor(ABC):

    @abstractmethod
    def pay(self, amount: float) -> None:
        pass


"""
===============================
2. Concrete Implementations (SRP)
===============================
"""
class StripePaymentProcessor(PaymentProcessor):

    logger = logging.getLogger("StripePaymentProcessor")
    PAYMENT_MESSAGE = "Paid $ %s using Stripe"

    def pay(self, amount: float) -> None:
        if self.logger.isEnabledFor(logging.INFO):
            self.logger.info(self.PAYMENT_MESSAGE % amount)


class RazorpayPaymentProcessor(PaymentProcessor):

    logger = logging.getLogger("RazorpayPaymentProcessor")
    PAYMENT_MESSAGE = "Paid $ %s using Razorpay"

    def pay(self, amount: float) -> None:
        if self.logger.isEnabledFor(logging.INFO):
            self.logger.info(self.PAYMENT_MESSAGE % amount)


class PayPalPaymentProcessor(PaymentProcessor):

    logger = logging.getLogger("PayPalPaymentProcessor")
    PAYMENT_MESSAGE = "Paid $%s using PayPal"

    def pay(self, amount: float) -> None:
        if self.logger.isEnabledFor(logging.INFO):
            self.logger.info(self.PAYMENT_MESSAGE % amount)


class PaymentProvider(Enum):
    STRIPE = "STRIPE"
    RAZORPAY = "RAZORPAY"
    PAYPAL = "PAYPAL"


"""
===============================
3. Factory (OCP)
===============================
"""
class PaymentProcessorFactory:

    def __init__(self, processors: Dict[PaymentProvider, PaymentProcessor]):
        self.processors = processors

    def get(self, provider: PaymentProvider) -> PaymentProcessor:

        if provider is None:
            raise ValueError("Payment provider must be specified")

        processor = self.processors.get(provider)

        if processor is None:
            raise ValueError(f"Unsupported payment provider: {provider}")

        return processor


"""
===============================
4. Business Logic (SRP + DIP)
===============================
"""
class CheckoutService:

    logger = logging.getLogger("CheckoutService")

    def __init__(self, factory: PaymentProcessorFactory):
        self.factory = factory

    """
    Executes checkout using selected payment provider.

    :param amount: payment amount (must be > 0)
    :param provider: payment provider name
    """
    def checkout(self, amount: float, provider: PaymentProvider) -> None:

        if amount <= 0:
            raise ValueError("Amount must be greater than zero")

        if self.logger.isEnabledFor(logging.INFO):
            self.logger.info("Starting checkout with provider: %s", provider.name)

        processor = self.factory.get(provider)
        processor.pay(amount)


class CheckoutServiceMainApplication:

    @staticmethod
    def main():
        factory = PaymentProcessorFactory({
            PaymentProvider.STRIPE: StripePaymentProcessor(),
            PaymentProvider.RAZORPAY: RazorpayPaymentProcessor(),
            PaymentProvider.PAYPAL: PayPalPaymentProcessor()
        })

        checkout_service = CheckoutService(factory)
        checkout_service.checkout(1500.0, PaymentProvider.STRIPE)


if __name__ == "__main__":
    CheckoutServiceMainApplication.main()


"""
===============================
Output
===============================

s2tha@thalasi-windows MINGW64 /d/git/low-level-design-practices/codeexamples/solid/better (main)
$ python3 CheckoutServiceMainApplication.py
INFO:CheckoutService:Starting checkout with provider: STRIPE
INFO:StripePaymentProcessor:Paid $ 1500.0 using Stripe

s2tha@thalasi-windows MINGW64 /d/git/low-level-design-practices/codeexamples/solid/better (main)
$

===============================
5. SOLID Principles
===============================

| Principle | Status | Why |
| ----------- | ------ | --------------------------------------- |
| S – SRP | Y | Checkout ≠ provider creation |
| O – OCP | Y | Add provider without modifying Checkout |
| L – LSP | Y | All processors interchangeable |
| I – ISP | Y | Small focused interface |
| D – DIP | Y | Depends on abstraction |
"""